homeassistant/components/system_log.py

```diff
@@ -0,0 +1,121 @@
+"""
+Support for system log.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/system_log/
+"""
+import os
+import re
+import asyncio
+import traceback
+from io import StringIO
+
+import voluptuous as vol
+
+from homeassistant.bootstrap import DATA_SYSTEM_LOG
+from homeassistant.config import load_yaml_config_file
+import homeassistant.helpers.config_validation as cv
+from homeassistant.components.http import HomeAssistantView
+
+DOMAIN = 'system_log'
+DEPENDENCIES = ['http']
+SERVICE_CLEAR = 'clear'
+
+CONF_MAX_ENTRIES = 'max_entries'
+
+DEFAULT_MAX_ENTRIES = 50
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Optional(CONF_MAX_ENTRIES,
+                     default=DEFAULT_MAX_ENTRIES): cv.positive_int,
+    }),
+}, extra=vol.ALLOW_EXTRA)
+
+SERVICE_CLEAR_SCHEMA = vol.Schema({})
+
+
+@asyncio.coroutine
+def async_setup(hass, config):
+    """Set up the logger component."""
+    conf = config[DOMAIN]
+
+    handler = hass.data[DATA_SYSTEM_LOG]
+    handler.resize(conf.get(CONF_MAX_ENTRIES))
+
+    hass.http.register_view(AllErrorsView(handler))
+    yield from hass.components.frontend.async_register_built_in_panel(
+        'system-log', 'System Log', 'mdi:monitor')
+
+    @asyncio.coroutine
+    def async_service_handler(service):
+        """Handle logger services."""
+        # Only one service so far
+        handler.records.clear()
+
+    descriptions = yield from hass.async_add_job(
+        load_yaml_config_file, os.path.join(
+            os.path.dirname(__file__), 'services.yaml'))
+
+    hass.services.async_register(
+        DOMAIN, SERVICE_CLEAR, async_service_handler,
+        descriptions[DOMAIN].get(SERVICE_CLEAR),
+        schema=SERVICE_CLEAR_SCHEMA)
+
+    return True
+
+
+def _figure_out_source(record):
+    # If a stack trace exists, extract filenames from the entire call stack.
+    # The other case is when a regular "log" is made (without an attached
+    # exception). In that case, just use the file where the log was made from.
+    if record.exc_info:
+        stack = list(map(lambda x: x[0],
```

Comment:
Use a list comprehension instead of `map` and `lambda`.

Reply:
I will fix that! 👍